src/lib/core/TLVReader.cpp

```diff
@@ -576,20 +566,19 @@ CHIP_ERROR TLVReader::VerifyEndOfContainer()
 
 CHIP_ERROR TLVReader::Next()
 {
-    CHIP_ERROR err;
-    TLVElementType elemType = ElementType();
+    ReturnErrorOnFailure(Skip());
+    ReturnErrorOnFailure(ReadElement());
 
-    err = Skip();
-    if (err != CHIP_NO_ERROR)
-        return err;
+    TLVElementType elemType = ElementType();
 
-    err = ReadElement();
-    if (err != CHIP_NO_ERROR)
-        return err;
+    VerifyOrReturnError(elemType != TLVElementType::EndOfContainer, CHIP_END_OF_TLV);
 
-    elemType = ElementType();
-    if (elemType == TLVElementType::EndOfContainer)
-        return CHIP_END_OF_TLV;
+    // Ensure that GetDataPtr calls can be called immediately after Next, so
+    // that `Get(ByteSpan&)` does not need to advance buffers and just works
+    if (TLVTypeIsString(elemType) && (GetLength() != 0))
+    {
+        ReturnErrorOnFailure(EnsureData(CHIP_ERROR_TLV_UNDERRUN));
```

Comment:
This _should_ be ok, in a sane world, but we've had issues in the past with shifts in where EnsureData() happens, so be ready for random regressions....